src/controllers/auth-controller.ts

```diff
@@ -75,32 +90,51 @@ export default class AuthController extends AbstractController {
       }
       await this.tokens.delete(closeAllSessions ? { userId } : { token: refreshTokenFromBody });
       await addAudit('logout', '', null, userId);
-      return res.send({ ok: true });
+      return { ok: true };
     } catch (e) {
       // README goes first as it is subclass of JsonWebTokenError
       if (e instanceof TokenExpiredError) {
-        return next(new CustomError('Token expired', 401));
+        next(new CustomError('Token expired', 401));
       }
       if (e instanceof JsonWebTokenError) {
-        return next(new CustomError('Token invalid', 401));
+        next(new CustomError('Token invalid', 401));
       }
-      return next(e);
+      next(e);
     }
-  };
+  }
 
-  private login = async (req: Request, res: Response, next: NextFunction) => {
-    try {
-      const user: User = (await authenticateLocal(req, res)) as User;
-      const { token, refreshToken } = signTokens({ userId: user.id, userRole: user.role });
-      await this.tokens.save(new RefreshToken(refreshToken, user.id));
-      return res.json({ user, token, refreshToken });
-    } catch (e) {
-      return next(e);
-    }
-  };
+  // TODO: move error handling to separate layer
+  /**
+   * Login to application.
+   * @param {WithCredentials} _userCreds User credentials
+   */
+  @POST
+  @Path('/login')
+  @Security('*', 'local')
+  @Response<SuccessAuthDto>(200, 'Successfully logged in.')
+  @Response<CustomError>(401, 'Authentication failed.')
+  public async login(_userCreds: WithCredentials, @ContextRequest req: Request): Promise<SuccessAuthDto> {
+    // @ts-ignore FIXME
+    const { user } = req;
+    const { token, refreshToken } = signTokens({ userId: user.id, userRole: user.role });
+    await this.tokens.save(new RefreshToken(refreshToken, user.id));
+    return { user, token, refreshToken };
+  }
 
-  private refresh = async (req: Request, res: Response, next: NextFunction) => {
-    const refreshTokenFromBody: string = req.body.refreshToken;
+  // TODO: move error handling to separate layer
```

Comment:
Can you tell me more about what you mean by?

Reply:
Separate layer to match error with http error code